numba/tests/test_record_dtype.py

```diff
@@ -1371,23 +1368,260 @@ def foo(rec, flag=None):
         self.assertEqual(foo(self.ab_rec1, flag=1), self.ab_rec1[0] + k + 20)
 
 
-class TestRecordArrayExceptions(TestCase):
+class TestNestedArrays(TestCase):
+
+    def get_cfunc(self, pyfunc, argspec):
+        cres = compile_isolated(pyfunc, argspec)
+        return cres.entry_point
+
+    def test_record_write_array(self):
+        # Testing writing to a 1D array within a structured type
+
+        nbval = np.recarray(1, dtype=recordwitharray)
+        nbrecord = numpy_support.from_dtype(recordwitharray)
+        cfunc = self.get_cfunc(record_write_array, (nbrecord,))
+        cfunc(nbval[0])
+
+        expected = np.recarray(1, dtype=recordwitharray)
+        expected[0].g = 2
+        expected[0].h[0] = 3.0
+        expected[0].h[1] = 4.0
+        np.testing.assert_equal(expected, nbval)
 
-    def test_nested_array_in_buffer_raises(self):
-        # see issue #6473
-        @njit()
-        def foo(x):
-            x["y"][0] = 1
+    def test_record_write_2d_array(self):
+        # Test writing to a 2D array within a structured type
 
-        dt = np.dtype([("y", (np.uint64, 5)),])
-        x = np.ones(1, dtype=dt)
-        with self.assertRaises(TypingError) as e:
-            foo(x)
-        ex1 = "The dtype of a Buffer type cannot itself be a Buffer type"
-        ex2 = "unsupported Buffer was: nestedarray(uint64, (5,))"
-        excstr = str(e.exception)
-        self.assertIn(ex1, excstr)
-        self.assertIn(ex2, excstr)
+        nbval = np.recarray(1, dtype=recordwith2darray)
+        nbrecord = numpy_support.from_dtype(recordwith2darray)
+        cfunc = self.get_cfunc(record_write_2d_array, (nbrecord,))
+        cfunc(nbval[0])
+
+        expected = np.recarray(1, dtype=recordwith2darray)
+        expected[0].i = 3
+        expected[0].j[:] = np.asarray([5.0, 6.0, 7.0, 8.0, 9.0, 10.0],
+                                      np.float32).reshape(3, 2)
```

Comment:
The expected value should ideally come from execution of the Python function being JIT compiled as that is what this JIT function is attempting to provide a native equivalent of. This comment applies throughout the proposed changes, however, in the interests of getting this functionality on mainline, unless @gmarkall feels strongly otherwise, we can leave it for now.

Reply:
I'm happy with it for now.

Reply:
fwiw, this code comes from `master`. it shows in the diff because I moved it.

Reply:
improvement here https://github.com/numba/numba/pull/7359/commits/be6369a07bc83d77de34d1d3425248cfe78c08e0